workflow/parser.py

```diff
@@ -0,0 +1,67 @@
+import os
+import subprocess
+
+
+class Parser:
+    def __init__(self, params):
+        self.params = params
+
+    def parse(self):
+
+        '''
+        - parse input config file 
+        - build TPS call 
+        - call TPS 
+        - gather outputs 
+        - return:
+            - list of outputs 
+            - output_folder
+            - out_label
+
+      
+        '''
+        # # grab TPS self.params 
+        build = []
+        OUT_FOLDER = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+        OUT_LABEL = ""
+
+        # add TPS run call as first arg
+        build.extend(["bash", "./scripts/run"])
+
+        # grab self.params from config
+        req = self.params[1]["TPS"][0]["required"]
```

Comment:
What is the `params[1]` indexing?  What else is in `params`?

Reply:
params is a complex nested dictionary structure due to the way the config file is set up (using "-" to denote values). We can change this in order to made indexing the config file easier but as long as the structure doesn't change (by a user), there shouldn't be any issues.

Reply:
I think I understand.  params is a list with values that are three dictionaries, and each of those dictionaries has a single key (Cytoscape, TPS, or Annoations)?  Can the outermost layer be a dictionary instead of a list?

Reply:
I think switching the format will make it easier to debug. Here is an article about yaml (I think you maybe shared this with me earlier) https://hackersandslackers.com/simplify-your-python-projects-configuration/. It introduced the library `confuse` that might be useful to use when parsing. It offers functionality to do param checking such as `as_int() as_filename()` etc.

Reply:
https://confuse.readthedocs.io/en/latest/usage.html